### PR TITLE
Make accessing dunder-names on a CTrait object an AttributeError

### LINKED ISSUE #1463: non-`HasTrait` classes with attributes of type `CTrait` break Sphinx + Napoleon autodoc
When building the documentation with Sphinx 4.0.1, we get this output. (See for example this [GitHub Actions run](https://github.com/enthought/traits/pull/1461/checks?check_run_id=2568557545) for #1461.)

```
Warning, treated as error:
autodoc: failed to determine <traits.ctrait.CTrait object at 0x7f74367d6220> to be documented, the following exception was raised:
Handler <function _skip_member at 0x7f7436e64280> for event 'autodoc-skip-member' threw an exception (exception: 'NoneType' object has no attribute 'rpartition')
```

To get CI up and running again, I've added a restriction to the Sphinx version in #1462. But sooner or later we'll want to remove that restriction.

### PR DESCRIPTION
**Pull request by mdickinson**
This PR changes attribute lookups on `CTrait` objects to raise `AttributeError` for names beginning and ending with a double underscore, and for which the normal attribute lookup rules don't return anything. Previously, those attribute lookups would return `None`, and this broke napoleon, which expected that a lookup of `__qualname__` on a `CTrait` object would either raise or return a string.

Attribute lookups for names that _don't_ begin and end with a double underscore are unchanged.

No documentation  update: it's not currently documented that arbitrary attribute accesses on a `CTrait` object return `None`, and I'm not sure that I really want it to be - it's not a "feature" of Traits that I'd want to encourage people to rely on.

Fixes #1463.

**Checklist**
- [x] Tests
- [ ] Update API reference (`docs/source/traits_api_reference`)
- [ ] Update User manual (`docs/source/traits_user_manual`)
- [ ] Update type annotation hints in `traits-stubs`

### CODE REVIEW

File: traits/ctraits.c [hunk 1]
```diff
@@ -2974,18 +2974,63 @@ trait_traverse(trait_object *trait, visitproc visit, void *arg)
 |  Handles the 'getattr' operation on a 'CTrait' instance:
 +----------------------------------------------------------------------------*/
 
+/* Return 1 if name starts and ends with "__", else 0. Return -1 with
+   an exception set on failure. */
+
+static int
+is_dunder_name(PyObject *name) {
```
Comment (mdickinson) [inline traits/ctraits.c:2981]: Style nit; opening brace should be on the next line.

File: traits/ctraits.c [hunk 2]
```diff
@@ -2974,18 +2974,63 @@ trait_traverse(trait_object *trait, visitproc visit, void *arg)
 |  Handles the 'getattr' operation on a 'CTrait' instance:
 +----------------------------------------------------------------------------*/
```
Comment (mdickinson) [inline traits/ctraits.c:2975]: This comment is now out of place; it needs to be moved so  that it's back above the definition of `trait_get_attro`.

File: traits/ctraits.c (hunk 1 above)
Comment (mdickinson) [reply traits/ctraits.c]: Fixed.

File: traits/ctraits.c [hunk 3]
```diff
@@ -2970,22 +2970,71 @@ trait_traverse(trait_object *trait, visitproc visit, void *arg)
...
+       exception to propagate. */
+
+    is_dunder = is_dunder_name(name);
+    if (is_dunder < 0) {
+        return NULL;
+    }
+    else if (is_dunder) {
+        /* Retry the attribute lookup so that we get the correct exception
+           message. */
+        return PyObject_GenericGetAttr((PyObject *)obj, name);
```
Comment (rahulporuri) [inline traits/ctraits.c:3032]: IIUC, we're having to retry in this clause instead of simply removing the `PyErr_Clear` outside of the conditional statements because the other two clauses still need/want to use `PyErr_Clear`. Is that correct?

File: traits/ctraits.c [hunk 4]
```diff
@@ -2970,22 +2970,71 @@ trait_traverse(trait_object *trait, visitproc visit, void *arg)
...
-    return Py_None;
+    /* The attribute lookup failed. For backwards compatibility, we return
+       `None` in this case, *except* for attribute names that start and end
+       with a double underscore. Those attribute names may have special meaning
+       to the interpreter, and `None` may not be a valid value, so we allow the
+       exception to propagate. */
+
+    is_dunder = is_dunder_name(name);
+    if (is_dunder < 0) {
+        return NULL;
```
Comment (rahulporuri) [inline traits/ctraits.c:3027]: Are we being defensive here when we handle the case where `name` isn't in the "canonical" representation? I'm not even sure what "canonical" representation means. Does it mean "invalid" unicode?

File: traits/tests/test_ctraits.py [hunk 5]
```diff
@@ -308,8 +308,33 @@ def _bar_changed(self, new):
...
+            "__a__b_",
+            "_",
+        ]
+
+        dunder_names = [
+            "__package__",
+            "__a__",
+            "____",
+            "___",
+            "__",
```
Comment (rahulporuri) [inline traits/tests/test_ctraits.py:326]: should `__`, `___` and sisters (integer multiples of `_` in between two other `_`) really be considered as dundernames in Python? I assumed that dundernames were `__X__` where x can be an alphanumeric string of any length..

**Review by rahulporuri — APPROVED** [review]
Mostly LGTM but with a few questions

References I used when reviewing this PR 
- https://docs.python.org/3.6/c-api/unicode.html?highlight=pyunicode_ready#c.PyUnicode_READY (this is why `is_dunder_name` will set an exception on failure)
- https://docs.python.org/3.6/c-api/unicode.html?highlight=pyunicode_ready#c.PyUnicode_GET_LENGTH
- https://docs.python.org/3.6/c-api/unicode.html?highlight=pyunicode_ready#c.PyUnicode_KIND (returns one of wchar, 1 byte, 2 byte and 4 byte)
- https://docs.python.org/3.6/c-api/unicode.html?highlight=pyunicode_ready#c.PyUnicode_DATA
- https://docs.python.org/3.6/c-api/unicode.html?highlight=pyunicode_ready#c.PyUnicode_READ
- https://docs.python.org/3.6/c-api/none.html?highlight=py_return_none#c.Py_RETURN_NONE (replaces `Py_INCREF(Py_None); return Py_None`)

File: traits/tests/test_ctraits.py (hunk 5 above)
Comment (mdickinson) [reply traits/tests/test_ctraits.py:326]: Yes, there are some corner cases here that aren't clear. I went with the equivalent of `my_string.startswith("__") and my_string.endswith("__")` because it's easy to describe, understand and implement, and because it matches CPython's own choices with respect to string mangling of class variable names (`__a__` won't be mangled because it's a dunder name; `__a` _will_ be mangled because it's not, and for whatever reason, `__` is _not_ mangled, so it's treated like the dunder names).

In the worst case, we'll encounter a use of `__` that really needs lookups to return `None`. I doubt that will happen, and if it does turn out to be a real need we can update the code accordingly.

File: traits/ctraits.c (hunk 4 above)
Comment (mdickinson) [reply traits/ctraits.c:3027]: Yes, we're being defensive, possibly unnecessarily so. 

For background on what canonical means, see: [PEP 393](https://www.python.org/dev/peps/pep-0393). Essentially, some deprecated, legacy APIs can produce `Py_Unicode` instances that don't have all of the expected fields initialised, but most of the Py_Unicode C-API expects those fields to be initialised and valid (i.e., they expect the object to be "canonical"). By Python 3.12 the legacy APIs will go away and this issue disappears: we can assume that any `Py_Unicode` object is canonical and drop the `Py_UnicodeReady` call.

This branch is here because `Py_UnicodeReady` itself can fail, as a result of memory allocation errors. It would be very hard to ever exercise this branch, but it should still be there.

Comment (mdickinson) [reply traits/ctraits.c:3027]: So to be clear, this branch isn't simply handling the case that `name` is not canonical: it's handling the case where `name` is not canonical *and* the attempt to make it canonical via `PyUnicode_Ready` failed. In practice, that should be a very rare occurrence indeed.

File: traits/ctraits.c (hunk 3 above)
Comment (mdickinson) [reply traits/ctraits.c:3032]: Hmm. This is really just laziness. What I want to do here is re-raise the _original_ `AttributeError`. I don't want to create my own exception message, since the one from the interpreter is likely to be better (and may even change with more recent versions of Python - there's a lot of work on error messages right now).

I can't simply remove the `PyErr_Clear` because `is_dunder_name` might set an exception, and it's an error to set another exception if there's already an exception set (or at least, I can't find any evidence that it's okay to do so, and there's CPython core code that goes out of its way to avoid this).

The right way to handle this without the retry is to do a `PyErr_Fetch` / `PyErr_Restore` pairing. I'll aim to update the PR to do that.

Comment (mdickinson) [reply traits/ctraits.c:3032]: > and it's an error to set another exception if there's already an exception set (or at least, I can't find any evidence that it's okay to do so, and there's CPython core code that goes out of its way to avoid this).

Hmm, actually, looking into the source, this may be okay. Investigating ...

**Comment by rahulporuri** [comment]
Still LGTM

File: traits/ctraits.c (hunk 3 above)
Comment (mdickinson) [reply traits/ctraits.c]: From the source, all the exception setting mechanisms end up at `_PyErr_SetObject`, which, if an exception is already set, applies implicit exception chaining. I think that's good enough for this case. I've reworked the code accordingly.

**Comment by mdickinson** [comment]
> Still LGTM

Sorry, @rahulporuri: I think the above message came just a couple of minutes before I updated to simplify the exception handling. Could you do another quick check?

**Review by rahulporuri — APPROVED** [review]
Still still LGTM 🎉

File: traits/ctraits.c (hunk 3 above)
Comment (rahulporuri) [reply traits/ctraits.c]: > From the source

I didn' mean to send you down a rabbit hole 😅

**Comment by mdickinson** [comment]
@rahulporuri Thank you! Merging.
